### PR TITLE
Allow closure to unsafe fn coercion

### LINKED ISSUE #57883: Direct coercion from closures to `unsafe fn` pointers
While you can coerce closures to `fn` pointers, it is currently not possible to coerce closures to `unsafe fn` pointers since that requires a transitive coercion from `fn` pointers to `unsafe fn` pointers. We do not allow transitive coercions. Example:
```rust
static A: fn() = || {}; // OK
static B: unsafe fn() = A; // OK
static C: unsafe fn() = || {}; // ERROR; would require the two coercions above transitively.
```
This behavior of current Rust was unexpected in https://github.com/rust-lang/rfcs/pull/2592#discussion_r250462675.

@cramertj had the idea to implement an immediately attainable direct coercion between closures and `unsafe fn` pointers rather than go the more complex and contentious route of transitive coercions (which may have unforeseen consequences). This idea seems reasonable to me.

cc @nikomatsakis @arielb1 

PS: should we fcp the PR once made?

### PR DESCRIPTION
**Pull request by taiki-e**
Closes #57883

### CODE REVIEW

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -2441,7 +2441,7 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     /// type with the same signature. Detuples and so forth -- so
     /// e.g., if we have a sig with `Fn<(u32, i32)>` then you would get
     /// a `fn(u32, i32)`.
```
Comment (Centril) [inline src/librustc/ty/context.rs]: Can you extend the doc-comment with note about the role of `unsafety`? Notably, the closure itself should not be able to do unsafe stuff.

File: src/librustc/mir/mod.rs [hunk 2]
```diff
@@ -2248,7 +2248,7 @@ pub enum CastKind {
     ReifyFnPointer,
 
     /// Converts non capturing closure to fn()
-    ClosureFnPointer,
```
Comment (Centril) [inline src/librustc/mir/mod.rs]: Can you extend doc-comment here as well? Same aspect about what the body of the closure can and cannot do would be good info.

File: src/librustc_typeck/check/coercion.rs [hunk 3]
```diff
@@ -714,16 +714,17 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
 
         let hir_id_a = self.tcx.hir().as_local_hir_id(def_id_a).unwrap();
         match b.sty {
-            ty::FnPtr(_) if self.tcx.with_freevars(hir_id_a, |v| v.is_empty()) => {
+            ty::FnPtr(fn_ty) if self.tcx.with_freevars(hir_id_a, |v| v.is_empty()) => {
                 // We coerce the closure, which has fn type
                 //     `extern "rust-call" fn((arg0,arg1,...)) -> _`
                 // to
                 //     `fn(arg0,arg1,...) -> _`
```
Comment (Centril) [inline src/librustc_typeck/check/coercion.rs]: Comment needs an update re. the possibility of `unsafe fn(arg0, arg1, ...)) -> _`.

File: src/test/run-pass/typeck-closure-to-unsafe-fn-ptr.rs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+unsafe fn call_unsafe(func: unsafe fn() -> ()) -> () {
+    func()
+}
+
+pub fn main() {
+    unsafe { call_unsafe(|| {}); }
```
Comment (Centril) [inline src/test/run-pass/typeck-closure-to-unsafe-fn-ptr.rs]: Can you extend the tests (in another file, `// compile-fail`) checking that:
```rust
let _: unsafe fn() = || /* do random stuff that requires `unsafe { ... }` */;
```
does not compile?

**Review by Centril — CHANGES_REQUESTED** [review]
👍

**Review by Centril — APPROVED** [review]

**Comment by oli-obk** [comment]
Impl lgtm. Issue suggests that this doesn't need an FCP.

@bors r+

I opened an issue (https://github.com/rust-lang/rust/issues/59588) talking about my unhappiness with the fact that this PR had to touch three enums. Note that this is absolutely not this PR's fault, but just the way our datastructes are laid out.

**Pull request merged**